src/media/media.ts

```diff
@@ -0,0 +1,820 @@
+   *    2) It will retry media with system defaults and no other options
+   *        (such as `deviceId`s, `videoFrameRate`, & `videoResolution`)
+   * 
+   * Warning: if `sdk.media.requestPermissions(type)` has NOT been called before 
+   *  calling `startMedia`, this function will call `sdk.media.requestPermissions(type)`.
+   *  If calling `startMedia` with both `audio` and `video` _before_ requesting permissions,
+   *  `startMedia` will attempt to gain permissions for `audio` first and then `video` (because
+   *  media permissions must be requested one at a time). If `audio` fails, it will 
+   *  not attempt to gain permissions for `video` – the error will stop execution. 
+   * 
+   * @param opts video and/or audio default device or deviceId. Defaults to `{video: true, audio: true}`
+   * @param retry whether the sdk should retry on an error
+   */
+  async startMedia (opts: IMediaRequestOptions = { video: true, audio: true }, retry: boolean = true): Promise<MediaStream> {
+    /* this will set media type to `truthy` if `null` was passed in */
+    const constraints: any = this.getStandardConstraints(opts);
+    const conversationId = opts.session?.conversationId;
+    const sessionId = opts.session?.id;
+    const isFirefox = browserama.isFirefox;
+    const requestingVideo = opts.video || opts.video === null;
+    const requestingAudio = opts.audio || opts.audio === null;
+
+    const { micPermissionsRequested, cameraPermissionsRequested } = this.getState();
+    let mediaStreamFromPermissions: MediaStream;
+
+    /* if we haven't requested permissions for audio yet, we need to do that first */
+    if (requestingAudio && !micPermissionsRequested) {
+      const optsWithOutsession = opts;
+      delete opts.session;
+      this.sdk.logger.info('attempted to get audio media before permissions checked. requesting audio permissions first', {
+        opts: optsWithOutsession,
+        sessionId,
+        conversationId,
+      });
+      mediaStreamFromPermissions = await this.requestMediaPermissions('audio', true, opts) as MediaStream;
+    }
+
+    /* if we haven't requested permissions for video yet, we need to do that first */
+    if (requestingVideo && !cameraPermissionsRequested) {
+      const optsWithOutsession = opts;
+      delete opts.session;
+      this.sdk.logger.info('attempted to get video media before permissions checked. requesting video permissions first', {
+        opts: optsWithOutsession,
+        sessionId,
+        conversationId,
+      });
+      const videoStream = await this.requestMediaPermissions('video', true, opts) as MediaStream;
+
+      if (!mediaStreamFromPermissions) {
+        mediaStreamFromPermissions = videoStream;
+      } else {
+        videoStream.getTracks().forEach(t => mediaStreamFromPermissions.addTrack(t));
+      }
+    }
+
+    /* if we preserved media from the permissions check, go ahead and return it (because it came from `startMedia`) */
+    if (mediaStreamFromPermissions) {
```

Comment:
I know this is nitpick edge case, but what if they already have camera permissions, but not mic permissions. This would mean they got a mediaStreamFromPermissions for audio, but no video because we bailed out here. The only case I could think of this happening is if they did proactive on camera, but not mic, but I think we could rewrite this is a way that works regardless.

Is there a pitfall with doing something like this:

```
let stream = ...;
if (requestingAudio) {
  if (!micPermissionsRequested) {
    // get stream from permissions
  } else {
    // get stream regularly
  }
}

// similar to above for video but merge tracks if stream exists

return stream;
```

Reply:
Even if we have mic/camera permissions, i think we should gUM separately for them both. this allows falling back for an individual media rather than losing everything. I think it will also allow us more fine grained debugging/logging.

Reply:
I was thinking more about this last night and I could see a potential issue with always calling gUM separately. I'm still for it, but wanted to talk it through and leave a paper trail. 

Potential Con: 
* scenario: call `startMedia({video: true, audio: true})` and gUM is called with audio first and then video. What happens if audio fails? We would have to return the error for audio, but video would never even be tried. Even with the other scenario of audio passing but video failing. What would we return then? The caller would need to implement its own retry logic. 

We would just need to make sure we document it well if we stick with the gUM separately. And as I am writing this out, I think we would be ok to always throw an error even if one of the gUM media types succeeded (and make sure to clean up the media). This is actually pretty inline with the current browser API. If both media types are requested at the same time but only one fails, the caller receives the error but no media (even for the other media type that didn't necessarily fail). 

I just wish there was a more clear way for use to inform the caller with one failed. But I am not in favor of "customizing" the response media/error object..

Reply:
I'm not sure what the "best" way to handle this is either. For sure, I think we should fall back to defaults for an individual gUM if it fails. If mic fails both though, I almost think we should just bail out. If camera fails both, I think we should "allow" it, but log an error. Example:

`startMedia({video: 'aff65a76-a77a6b', audio: 'blahblah'})`
* try gUM for audio specifying blahblah as the device.
* gUM fails so try gUM for audio with defaults.
* gUM succeeds for audio
* gUM for video; succeeds.
* merge and return stream

similar for video ^.

If mic fails the default gUM, throw error and bail. If camera fails the default gUM, log error and continue. I like the idea of not retrying for defaults if the error is caused by permissions though, I think we should keep that, but still following the same pattern of continuing if it's just camera that failed. Make sense? Thoughts and opinions?